File: LeetCode/0560. Subarray Sum Equals K/Solution.java
/**
 * 92 / 92 test cases passed.
 * Runtime: 35 ms
 * Memory Usage: 44.9 MB 
 */
class Solution {
    public int subarraySum(int[] nums, int k) {
        HashMap<Integer, Integer> rec = new HashMap<>();
        rec.put(0, 1);
        int ans = 0, psum = 0;
        for (Integer num : nums) {
            psum += num;
            if (rec.containsKey(psum - k)) {
                ans += rec.get(psum - k);
            }
            rec.put(psum, rec.getOrDefault(psum, 0) + 1);
        }
        return ans;
    }
}

File: LeetCode/0560. Subarray Sum Equals K/solution.cpp
/**
 * 92 / 92 test cases passed.
 * Runtime: 48 ms
 * Memory Usage: 40.5 MB 
 */
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> rec;
        rec[0] = 1;
        int ans = 0, psum = 0;
        for (auto &num : nums) {
            psum += num;
            if (rec.count(psum - k)) {
                ans += rec[psum - k];
            }
            ++ rec[psum];
        }
        return ans;
    }
};
